File: src/api/library.ts
import { callCommand } from './client'
import type {
  Axiom,
  AxiomDraft,
  Character,
  CharacterDraft,
  Entry,
  EntryDraft,
  EntityRef,
  EventDraft,
  EventParticipantDraft,
  EventRecord,
  Relation,
  RelationDraft,
} from '@/types/library'

export function listEntries(projectId: string): Promise<Entry[]> {
  return callCommand<Entry[]>('list_entries', { projectId }).then((items) => items ?? [])
}

export function createEntry(draft: EntryDraft): Promise<Entry> {
  return callCommand<Entry>('create_entry', { draft })
}

export function updateEntry(id: string, draft: EntryDraft): Promise<Entry> {
  return callCommand<Entry>('update_entry', { id, draft })
}

export function deleteEntry(id: string): Promise<void> {
  return callCommand<void>('delete_entry', { id })
}

export function listCharacters(projectId: string): Promise<Character[]> {
  return callCommand<Character[]>('list_characters', { projectId }).then((items) => items ?? [])
}

export function createCharacter(draft: CharacterDraft): Promise<Character> {
  return callCommand<Character>('create_character', { draft })
}

export function updateCharacter(id: string, draft: CharacterDraft): Promise<Character> {
  return callCommand<Character>('update_character', { id, draft })
}

export function deleteCharacter(id: string): Promise<void> {
  return callCommand<void>('delete_character', { id })
}

export function listEvents(projectId: string): Promise<EventRecord[]> {
  return callCommand<EventRecord[]>('list_events', { projectId }).then((items) => items ?? [])
}

export function createEvent(
  draft: EventDraft,
  participants: EventParticipantDraft[] = [],
): Promise<EventRecord> {
  return callCommand<EventRecord>('create_event', { draft, participants })
}

export function updateEvent(
  id: string,
  draft: EventDraft,
  participants: EventParticipantDraft[] = [],
): Promise<EventRecord> {
  return callCommand<EventRecord>('update_event', { id, draft, participants })
}

export function deleteEvent(id: string): Promise<void> {
  return callCommand<void>('delete_event', { id })
}

export function searchAxioms(projectId: string, query: string): Promise<Axiom[]> {
  return callCommand<Axiom[]>('search_axioms', { projectId, query }).then((items) => items ?? [])
}

export function createAxiom(draft: AxiomDraft): Promise<Axiom> {
  return callCommand<Axiom>('create_axiom', { draft })
}

export function updateAxiom(id: string, draft: AxiomDraft): Promise<Axiom> {
  return callCommand<Axiom>('update_axiom', { id, draft })
}

export function deleteAxiom(id: string): Promise<void> {
  return callCommand<void>('delete_axiom', { id })
}

export function listBacklinks(target: EntityRef): Promise<Relation[]> {
  return callCommand<Relation[]>('list_backlinks', { target }).then((items) => items ?? [])
}

export function createRelation(draft: RelationDraft): Promise<Relation> {
  return callCommand<Relation>('create_relation', { draft })
}

export function listRelations(projectId: string): Promise<Relation[]> {
  return callCommand<Relation[]>('list_relations', { projectId }).then((items) => items ?? [])
}

export function updateRelation(id: string, draft: RelationDraft): Promise<Relation> {
  return callCommand<Relation>('update_relation', { id, draft })
}

export function deleteRelation(id: string): Promise<void> {
  return callCommand<void>('delete_relation', { id })
}

File: src-tauri/src/commands/library.rs
use tauri::State;

use crate::{
    db::app_state::AppState,
    domain::{
        axiom::{Axiom, AxiomDraft, AxiomRepository},
        character::{Character, CharacterDraft, CharacterRepository},
        entry::{Entry, EntryDraft, EntryRepository},
        event::{Event, EventDraft, EventParticipantDraft, EventRepository},
        relation::{EntityRef, Relation, RelationDraft, RelationRepository},
    },
};

#[tauri::command]
pub fn list_entries(state: State<'_, AppState>, project_id: String) -> Result<Vec<Entry>, String> {
    state.with_database(|connection| EntryRepository::new(connection).list_active(&project_id))
}

#[tauri::command]
pub fn create_entry(state: State<'_, AppState>, draft: EntryDraft) -> Result<Entry, String> {
    state.with_database(|connection| EntryRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn update_entry(
    state: State<'_, AppState>,
    id: String,
    draft: EntryDraft,
) -> Result<Entry, String> {
    state.with_database(|connection| EntryRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn delete_entry(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| EntryRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn list_characters(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<Vec<Character>, String> {
    state.with_database(|connection| CharacterRepository::new(connection).list_active(&project_id))
}

#[tauri::command]
pub fn create_character(
    state: State<'_, AppState>,
    draft: CharacterDraft,
) -> Result<Character, String> {
    state.with_database(|connection| CharacterRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn update_character(
    state: State<'_, AppState>,
    id: String,
    draft: CharacterDraft,
) -> Result<Character, String> {
    state.with_database(|connection| CharacterRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn delete_character(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| CharacterRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn list_events(state: State<'_, AppState>, project_id: String) -> Result<Vec<Event>, String> {
    state.with_database(|connection| EventRepository::new(connection).list_active(&project_id))
}

#[tauri::command]
pub fn create_event(
    state: State<'_, AppState>,
    draft: EventDraft,
    participants: Vec<EventParticipantDraft>,
) -> Result<Event, String> {
    state.with_database(|connection| EventRepository::new(connection).create(draft, participants))
}

#[tauri::command]
pub fn update_event(
    state: State<'_, AppState>,
    id: String,
    draft: EventDraft,
    participants: Vec<EventParticipantDraft>,
) -> Result<Event, String> {
    state.with_database(|connection| EventRepository::new(connection).update(&id, draft, participants))
}

#[tauri::command]
pub fn delete_event(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| EventRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn search_axioms(
    state: State<'_, AppState>,
    project_id: String,
    query: String,
) -> Result<Vec<Axiom>, String> {
    state.with_database(|connection| AxiomRepository::new(connection).search(&project_id, &query))
}

#[tauri::command]
pub fn create_axiom(state: State<'_, AppState>, draft: AxiomDraft) -> Result<Axiom, String> {
    state.with_database(|connection| AxiomRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn update_axiom(
    state: State<'_, AppState>,
    id: String,
    draft: AxiomDraft,
) -> Result<Axiom, String> {
    state.with_database(|connection| AxiomRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn delete_axiom(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| AxiomRepository::new(connection).soft_delete(&id))
}

#[tauri::command]
pub fn list_backlinks(
    state: State<'_, AppState>,
    target: EntityRef,
) -> Result<Vec<Relation>, String> {
    state.with_database(|connection| RelationRepository::new(connection).list_backlinks(&target))
}

#[tauri::command]
pub fn create_relation(
    state: State<'_, AppState>,
    draft: RelationDraft,
) -> Result<Relation, String> {
    state.with_database(|connection| RelationRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn list_relations(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<Vec<Relation>, String> {
    state.with_database(|connection| RelationRepository::new(connection).list_project(&project_id))
}

#[tauri::command]
pub fn update_relation(
    state: State<'_, AppState>,
    id: String,
    draft: RelationDraft,
) -> Result<Relation, String> {
    state.with_database(|connection| RelationRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn delete_relation(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| RelationRepository::new(connection).soft_delete(&id))
}
